lib/pages/home/pages/search/widgets/top_posts.dart

```diff
@@ -0,0 +1,268 @@
+import 'dart:async';
+import 'dart:convert';
+import 'package:Okuna/models/community.dart';
+import 'package:Okuna/models/post.dart';
+import 'package:Okuna/models/top_post.dart';
+import 'package:Okuna/models/top_posts_list.dart';
+import 'package:Okuna/provider.dart';
+import 'package:Okuna/services/localization.dart';
+import 'package:Okuna/services/navigation_service.dart';
+import 'package:Okuna/services/user.dart';
+import 'package:Okuna/widgets/icon.dart';
+import 'package:Okuna/widgets/icon_button.dart';
+import 'package:Okuna/widgets/post/post.dart';
+import 'package:Okuna/widgets/posts_stream/posts_stream.dart';
+import 'package:Okuna/widgets/theming/primary_accent_text.dart';
+import 'package:async/async.dart';
+import 'package:flutter/material.dart';
+
+class OBTopPosts extends StatefulWidget {
+  final OBTopPostsController controller;
+
+  OBTopPosts({
+    this.controller,
+  });
+
+  @override
+  State<StatefulWidget> createState() {
+    return OBTopPostsState();
+  }
+}
+
+class OBTopPostsState extends State<OBTopPosts> with AutomaticKeepAliveClientMixin {
+  UserService _userService;
+  LocalizationService _localizationService;
+  NavigationService _navigationService;
+  CancelableOperation _getTrendingPostsOperation;
+  OBPostsStreamController _obPostsStreamController;
+
+  bool _needsBootstrap;
+  List<TopPost> _currentTopPosts = [];
+  List<Post> _currentPosts = [];
+  List<int> _excludedCommunities = [];
+  List<Post> _topPostsCacheData = [];
+  int _topPostLastViewedId;
+
+  @override
+  void initState() {
+    super.initState();
+    _needsBootstrap = true;
+    _obPostsStreamController = OBPostsStreamController();
+    if (widget.controller != null) widget.controller.attach(this);
+  }
+
+  @override
+  void dispose() async {
+    super.dispose();
+    await _userService.setStoredTopPosts(_currentTopPosts);
+    if (_getTrendingPostsOperation != null) _getTrendingPostsOperation.cancel();
+  }
+
+  Future refresh() {
+    return _obPostsStreamController.refreshPosts();
+  }
+
+  void scrollToTop() {
+    _obPostsStreamController.scrollToTop();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    var openbookProvider = OpenbookProvider.of(context);
+    _userService = openbookProvider.userService;
+    _localizationService = openbookProvider.localizationService;
+    _navigationService = openbookProvider.navigationService;
+
+    if (_needsBootstrap) _bootstrap();
```

Comment:
Why not the usual if ! _bootstrap ? bootstrap and always return a the PostsStream widget

Reply:
The reason is the `refreshOnCreate` 
```
    if (_needsBootstrap) _bootstrap();
    if (!_needsBootstrap) {
      return OBPostsStream(
        streamIdentifier: 'explorePostsTab',
        refresher: _postsStreamRefresher,
        onScrollLoader: _postsStreamOnScrollLoader,
        controller: _obPostsStreamController,
        isTopPostsStream: true,
        initialPosts: _currentPosts,
        refreshOnCreate: _currentPosts.length == 0,
```
If we return the widget after calling the bootstrap func (which fetches the cache), the OBPostStream is always initialised with no posts, this causes it to start refreshing content from the server resulting in the cache being ignored.  Ill give it another go 🤔

Reply:
Yeah, the initialPosts is not meant to be dynamic. If you want to listen for these final attributes changes you have to listen to the didUpdateWidget event.

Alright. So it's while it bootstraps the initial posts basically.